pkg/controller/component/lifecycle/kbagent.go

```diff
@@ -51,74 +56,160 @@ type kbagent struct {
 
-func (a *kbagent) Switchover(ctx context.Context, cli client.Reader, opts *Options) error {
-	la := &switchover{}
-	return a.checkedCallAction(ctx, cli, a.lifecycleActions.Switchover, la, opts)
+func (a *kbagent) Switchover(ctx context.Context, cli client.Reader, opts *Options, candidate string) error {
+	lfa := &switchover{
+		namespace:   a.namespace,
+		clusterName: a.clusterName,
+		compName:    a.compName,
+		roles:       a.roles,
+		candidate:   candidate,
+	}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.Switchover, lfa, opts)
 }
 
 func (a *kbagent) MemberJoin(ctx context.Context, cli client.Reader, opts *Options) error {
-	la := &memberJoin{
-		synthesizedComp: a.synthesizedComp,
-		pod:             a.pod,
+	lfa := &memberJoin{
+		namespace:   a.namespace,
+		clusterName: a.clusterName,
+		compName:    a.compName,
+		pod:         a.pod,
 	}
-	return a.checkedCallAction(ctx, cli, a.lifecycleActions.MemberJoin, la, opts)
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.MemberJoin, lfa, opts)
 }
 
 func (a *kbagent) MemberLeave(ctx context.Context, cli client.Reader, opts *Options) error {
-	la := &memberLeave{
-		synthesizedComp: a.synthesizedComp,
-		pod:             a.pod,
+	lfa := &memberLeave{
+		namespace:   a.namespace,
+		clusterName: a.clusterName,
+		compName:    a.compName,
+		pod:         a.pod,
 	}
-	return a.checkedCallAction(ctx, cli, a.lifecycleActions.MemberLeave, la, opts)
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.MemberLeave, lfa, opts)
 }
 
 func (a *kbagent) DataDump(ctx context.Context, cli client.Reader, opts *Options) error {
-	la := &dataDump{}
-	return a.checkedCallAction(ctx, cli, a.lifecycleActions.DataDump, la, opts)
+	lfa := &dataDump{}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.DataDump, lfa, opts)
 }
 
 func (a *kbagent) DataLoad(ctx context.Context, cli client.Reader, opts *Options) error {
-	la := &dataLoad{}
-	return a.checkedCallAction(ctx, cli, a.lifecycleActions.DataLoad, la, opts)
+	lfa := &dataLoad{}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.DataLoad, lfa, opts)
+}
+
+func (a *kbagent) AccountProvision(ctx context.Context, cli client.Reader, opts *Options, statement, user, password string) error {
+	lfa := &accountProvision{
+		statement: statement,
+		user:      user,
+		password:  password,
+	}
+	return a.checkedCallAction(ctx, cli, a.lifecycleActions.AccountProvision, lfa, opts)
+}
+
+func (a *kbagent) checkedCallAction(ctx context.Context, cli client.Reader, spec *appsv1alpha1.Action, lfa lifecycleAction, opts *Options) error {
+	if spec == nil || spec.Exec == nil {
+		return errors.Wrap(ErrActionNotDefined, lfa.name())
+	}
+	if err := a.precondition(ctx, cli, spec); err != nil {
+		return err
+	}
+	// TODO: exactly once
+	return a.callAction(ctx, cli, spec, lfa, opts)
+}
+
+func (a *kbagent) precondition(ctx context.Context, cli client.Reader, spec *appsv1alpha1.Action) error {
+	if spec.PreCondition == nil {
+		return nil
+	}
```

Comment:
It seems:
- if we set precondition of PreTerminate action to 'ClusterReady', it may block a unhealthy component be deleted.
- What's more, if we set precondition of Switchover action to 'ClusterReady' or 'ComponentReady', it may fail to do switchover if  the primary Pod's status is not running.
 
right?

Reply:
Yeah, so the precondition option is not suitable for all actions to set in practice.